signalfx-java/src/main/java/com/signalfx/signalflow/WebSocketTransport.java

```diff
@@ -0,0 +1,592 @@
+            }
+
+            if (code != 1000) {
+                this.error = new SignalFlowException(code, reason);
+                log.info("Lost WebSocket connection with {} ({}).", connection, code);
+
+                SignalFlowExceptionStreamMessage errorMessage = new SignalFlowExceptionStreamMessage(
+                        this.error);
+                for (TransportChannel channel : this.channels.values()) {
+                    channel.offer(errorMessage);
+                }
+            }
+
+            this.channels.clear();
+            this.connection = null;
+        }
+
+        @Override
+        public void onOpen(Connection connection) {
+            if (log.isDebugEnabled()) {
+                log.debug("open connection: " + connection);
+            }
+            this.connection = connection;
+
+            Map<String, String> authRequest = new HashMap<String, String>();
+            authRequest.put("type", "authenticate");
+            authRequest.put("token", this.token);
+
+            sendMessage(authRequest);
+        }
+
+        @Override
+        public void onMessage(byte[] data, int offset, int length) {
+
+            try {
+                byte[] messageBytes = Arrays.copyOfRange(data, offset, offset + length);
+                TransportDataMessage dataMessage = new TransportDataMessage(messageBytes);
+
+                if (dataMessage.getMessageType() == 5) {
+                    LinkedHashMap<String, Object> dataMap = new LinkedHashMap<String, Object>();
+                    dataMap.put("logicalTimestampMs", dataMessage.getLogicalTimestampMs());
+                    dataMap.put("data", dataMessage.getData());
+
+                    TransportChannel channel = channels.get(dataMessage.getChannelName());
+                    if ((channel != null) && (!channel.isClosed())) {
+                        StreamMessage streamMessage = new StreamMessage("data", null,
+                                objectMapper.writeValueAsString(dataMap));
+                        channel.offer(streamMessage);
+                    } else {
+                        log.debug("ignoring message. channel not found {}",
+                                dataMessage.getChannelName());
+                    }
+                }
+            } catch (JsonProcessingException ex) {
+                log.error("failed to process messages", ex);
+            }
+        }
+
+        @Override
+        public void onMessage(String data) {
+
+            try {
+                @SuppressWarnings("unchecked")
+                Map<String, Object> dataMap = objectMapper.readValue(data, Map.class);
+
+                // Intercept KEEP_ALIVE messages
+                String event = (String) dataMap.get("event");
+                if ((event != null) && (event.equals("KEEP_ALIVE"))) {
+                    return;
+                }
+
+                event = (String) dataMap.get("type");
+                if (event != null) {
```

Comment:
once upon a time early returns like that were considered poor programming.
anyway .. done